File: scala/core/src/main/scala/blockchain/BlockSourcedState.scala
package blockchain

import blockchain.models.BlockId
import cats.data.Chain
import cats.implicits.*
import cats.effect.implicits.*
import cats.effect.{Async, Ref, Resource}
import cats.effect.std.{Mutex, Semaphore}

trait BlockSourcedState[F[_], State]:
  def useStateAt[U](blockId: BlockId)(f: State => F[U]): F[U]

object BlockSourcedState:
  def make[F[_]: Async, State](
      initialState: F[State],
      initialEventId: F[BlockId],
      applyEvent: (State, BlockId) => F[State],
      unapplyEvent: (State, BlockId) => F[State],
      parentChildTree: BlockIdTree[F],
      currentEventChanged: BlockId => F[Unit]
  ): Resource[F, BlockSourcedState[F, State]] =
    (
      Mutex[F],
      initialState.flatMap(Ref.of[F, State]),
      initialEventId.flatMap(Ref.of[F, BlockId])
    )
      .mapN((mutex, stateRef, idRef) =>
        new TreeBlockSourcedState[F, State](
          applyEvent,
          unapplyEvent,
          parentChildTree,
          mutex,
          stateRef,
          idRef,
          currentEventChanged
        )
      )
      .toResource

class TreeBlockSourcedState[F[_]: Async, State](
    applyEvent: (State, BlockId) => F[State],
    unapplyEvent: (State, BlockId) => F[State],
    parentChildTree: BlockIdTree[F],
    mutex: Mutex[F],
    currentStateRef: Ref[F, State],
    currentEventIdRef: Ref[F, BlockId],
    currentEventChanged: BlockId => F[Unit]
) extends BlockSourcedState[F, State]:
  def useStateAt[U](blockId: BlockId)(f: State => F[U]): F[U] =
    mutex.lock.surround(
      for {
        currentEventId <- currentEventIdRef.get
        state <-
          (currentEventId == blockId)
            .pure[F]
            .ifM(
              currentStateRef.get,
              for {
                (unapplyChain, applyChain) <- parentChildTree.findCommonAncestor(currentEventId, blockId)
                currentState <- currentStateRef.get
                stateAtCommonAncestor <- unapplyEvents(
                  unapplyChain.tail,
                  currentState,
                  unapplyChain.head
                )
                newState <- applyEvents(applyChain.tail, stateAtCommonAncestor)
              } yield newState
            )
        res <- f(state)
      } yield res
    )

  private def unapplyEvents(
      eventIds: Chain[BlockId],
      currentState: State,
      newEventId: BlockId
  ): F[State] =
    eventIds.zipWithIndex.reverse.foldLeftM(currentState) {
      case (state, (eventId, index)) =>
        Async[F].uncancelable(_ =>
          for {
            newState <- unapplyEvent(state, eventId)
            nextEventId = eventIds.get(index - 1).getOrElse(newEventId)
            _ <- (
              currentStateRef.set(newState),
              currentEventIdRef.set(nextEventId),
              currentEventChanged(nextEventId)
            ).tupled
          } yield newState
        )
    }

  private def applyEvents(
      eventIds: Chain[BlockId],
      currentState: State
  ): F[State] =
    eventIds.foldLeftM(currentState) { case (state, eventId) =>
      Async[F].uncancelable(_ =>
        for {
          newState <- applyEvent(state, eventId)
          _ <- (
            currentStateRef.set(newState),
            currentEventIdRef.set(eventId),
            currentEventChanged(eventId)
          ).tupled
        } yield newState
      )
    }

File: dart/protocol/lib/src/blockchain/block_sourced_state.dart
import 'package:giraffe_sdk/sdk.dart';
import 'package:mutex/mutex.dart';

import 'block_id_tree.dart';

abstract class BlockSourcedState<State> {
  Future<T> useStateAt<T>(BlockId blockId, Future<T> Function(State state) f);
}

class BlockSourcedStateImpl<State> extends BlockSourcedState<State> {
  final Future<State> Function(State, BlockId) applyBlock;
  final Future<State> Function(State, BlockId) unapplyBlock;
  final BlockIdTree blockIdTree;
  final Mutex mutex = Mutex();
  State state;
  BlockId currentId;
  final Future<void> Function(BlockId) blockIdChanged;

  BlockSourcedStateImpl(
      {required this.applyBlock,
      required this.unapplyBlock,
      required this.blockIdTree,
      required this.state,
      required this.currentId,
      required this.blockIdChanged});

  @override
  Future<T> useStateAt<T>(BlockId blockId, Future<T> Function(State state) f) =>
      mutex.protect(() async {
        if (currentId == blockId) {
          return f(state);
        }
        final (unapplyChain, applyChain) =
            await blockIdTree.findCommonAncestor(currentId, blockId);
        for (int i = unapplyChain.length - 1; i > 1; i--) {
          final id = unapplyChain[i];
          state = await unapplyBlock(state, id);
          currentId = unapplyChain[i - 1];
          blockIdChanged(currentId);
        }
        for (int i = 1; i < applyChain.length; i++) {
          final id = applyChain[i];
          state = await applyBlock(state, id);
          currentId = applyChain[i];
          blockIdChanged(currentId);
        }
        return f(state);
      });
}
